src/Subscriber/PaymentSubscriber.php

```diff
@@ -130,15 +142,41 @@ public static function getSubscribedEvents(): array
      */
     public function onContextTokenUpdate(SalesChannelContextSwitchEvent $event)
     {
-        if ($event->getRequestDataBag()->get('adyenStateData')) {
-            try {
-                $this->paymentStateDataService->insertPaymentStateData(
-                    $event->getSalesChannelContext()->getToken(),
-                    $event->getRequestDataBag()->get('adyenStateData'),
-                    $event->getRequestDataBag()->get('adyenOrigin')
+        //State data from the frontend
+        $stateData = $event->getRequestDataBag()->get('adyenStateData');
+
+        if ($stateData) {
+            //Convert the state data into an array
+            $stateDataArray = json_decode($stateData, true);
+
+            //Payment method selected from the Shopware methods form
+            $selectedPaymentMethod = $this->paymentMethodRepository->search(
+                (new Criteria())
+                    ->addFilter(new EqualsFilter('id', $event->getRequestDataBag()->get('paymentMethodId'))),
+                Context::createDefaultContext()
+            )->first();
+
+            $selectedPaymentMethodIsStoredPM =
+                $selectedPaymentMethod->getFormattedHandlerIdentifier() == 'handler_adyen_oneclickpaymentmethodhandler';
+
+            $stateDataIsStoredPM = !empty($stateDataArray["paymentMethod"]["storedPaymentMethodId"]);
+
+            //Only store the state data if it matches the selected PM
+            if ($stateDataIsStoredPM == $selectedPaymentMethodIsStoredPM) {
+                try {
+                    $this->paymentStateDataService->insertPaymentStateData(
+                        $event->getSalesChannelContext()->getToken(),
```

Comment:
Does it maybe a good idea to also insert the selected payment method (id, handler, code - one of these which identifies the payment method in shopware) so later we don't have to rely on the satte data type?

Reply:
I think so, it could be handy, just not on this subscriber as we always check the new oncoming state.data against the selected PM. Created a ticket to look into it PW-3633